Added bool param to service creation for torch to optionally disable …

cchadowitz-pf opened the pull request:
…applying a softmax to the model output (default enabled)

Note: not sure if `_module._softmax_preprocess` (see torchmodule.h) is the right place to store this value, happy to move it around if desired.

Bycob commented on src/dto/mllib.hpp:
```diff
@@ -176,6 +176,15 @@ namespace dd
+              "model "
+              "already applies a sigmoid, and this should be set to false";
+      }
+      DTO_FIELD(Boolean, softmax_postprocess) = true;
```
We might want a higher level name for this parameter. In your case it's multilabel, but maybe we don't need the softmax in other cases either. @beniz any opinion on this?

Bycob replied:
In caffe we had the `multi_label` option, which effectively disabled the softmax: https://github.com/jolibrain/deepdetect/blob/master/src/backends/caffe/caffeinputconns.h#L193. It was in the input connector, but I think you could put it in mllib, since this is also arguably a property of the model/of the output.

Note that the softmax layer is already not used in some cases (eg regression), in general we prefer to let the user chose the type of tasks they want to perform, and do all the dirty work of plugging layers together and post processing the output behind the scenes

cchadowitz-pf commented:
Prior to my new test, I had `service_predict_native_bw` failing when I ran `ut_torch` - is that expected if I'm not running with GPU support?
```[ RUN      ] torchapi.service_predict_native_bw
[2022-07-21 10:27:29.563] [imgserv] [error] service creation mllib bad param: GPU is not available, service could not be created
[2022-07-21 10:27:29.563] [api] [error] service not found: "imgserv"
joutstr={"status":{"code":404,"msg":"NotFound","dd_code":1002,"dd_msg":"Service Not Found"}}
/deepdetect/tests/ut-torchapi.cc:232: Failure
Expected equality of these values:
  200
  jd["status"]["code"]
    Which is: 16-byte object <94-01 00-00 00-00 00-00 00-00 00-00 00-00 F6-01>
[  FAILED  ] torchapi.service_predict_native_bw (4 ms)
```

cchadowitz-pf commented:
I've tried running the tests with the changes in [4d010a6](https://github.com/jolibrain/deepdetect/pull/1448/commits/4d010a6b78f3f54073c7d8c483f3f84ca7e9f754) and can't seem to get the param to actually affect the `predict()` call correctly. It's like `_multi_label` isn't actually saved/stored between the mllib initialization and making a predict call. I've tried using `"multi_label": true` in both the service create `mllib` as well as in the predict `mllib`:
```
TEST(torchapi, service_predict_no_softmax_postprocess)
{
  // create service
  JsonAPI japi;
  std::string sname = "imgserv";
  std::string jstr
      = "{\"mllib\":\"torch\",\"description\":\"resnet-50\",\"type\":"
        "\"supervised\",\"model\":{\"repository\":\""
        + incept_repo
        + "\"},\"parameters\":{\"input\":{\"connector\":\"image\",\"height\":"
          "224,\"width\":224,\"rgb\":true,\"scale\":0.0039},\"mllib\":{"
          "\"nclasses\":1000,\"multi_label\":true}}}";
  std::string joutstr = japi.jrender(japi.service_create(sname, jstr));
  ASSERT_EQ(created_str, joutstr);

  // predict
  std::string jpredictstr
      = "{\"service\":\"imgserv\",\"parameters\":{\"input\":{\"height\":224,"
        "\"width\":224},\"output\":{\"best\":10},\"mllib\":{\"multi_label\":true}},\"data\":"
        "[\""
        + incept_repo + "cat.jpg\"]}";

  joutstr = japi.jrender(japi.service_predict(jpredictstr));
  JDoc jd = JDoc();
  std::cout << "joutstr=" << joutstr << std::endl;
  jd.Parse<rapidjson::kParseNanAndInfFlag>(joutstr.c_str());
  ASSERT_TRUE(!jd.HasParseError());
  ASSERT_EQ(200, jd["status"]["code"]);
  ASSERT_TRUE(jd["body"]["predictions"].IsArray());
  ASSERT_EQ(jd["body"]["predictions"][0]["classes"].Size(), 10);

  ASSERT_GT(jd["body"]["predictions"][0]["classes"][0]["prob"].GetDouble(),
            10.0);
  double sum = 0.0;
  for (size_t i = 0; i < jd["body"]["predictions"][0]["classes"].Size(); i++) {
      sum += jd["body"]["predictions"][0]["classes"][i]["prob"].GetDouble();
  }
  ASSERT_GT(sum, 1.0); // Somewhat redundant but still validates the test case
}
```

cchadowitz-pf commented:
LGTM! Still getting a failure on `torchapi.service_predict_native_bw` due to running the tests without a GPU, I assume that's expected?

cchadowitz-pf commented:
One additional thing - I have a second torch model that includes a softmax as part of its own `forward()` method:
```
        InceptionResnetV2_Logits_Predictions = F.softmax(InceptionResnetV2_Logits_Logits_MatMul)
        return InceptionResnetV2_Logits_Predictions
```
Obviously if I set the upcoming API param `"multi_label": true` then everything should be fine, but I don't know if this kind of scenario would prompt a different name for the param, as this model isn't a multi-label model, it just already performs the softmax internally and doesn't need post-processing.

Bycob commented:
> 

Our current policy is to export all the models for a given task with the same input / output format, so that there is one unique path in dede. As torchvision models output logits, this was set as the default. Do you have any way to remove the softmax from your model?

Bycob commented:
> LGTM! Still getting a failure on `torchapi.service_predict_native_bw` due to running the tests without a GPU, I assume that's expected?

Seems like this test should either not require GPU or not be built on CPU-only dede.

Bycob commented:
As for the CI: it fails with issues unrelated to this PR's changes. We are fixing the problems, after that it should pass.

cchadowitz-pf commented:
> Our current policy is to export all the models for a given task with the same input / output format, so that there is one unique path in dede. As torchvision models output logits, this was set as the default. Do you have any way to remove the softmax from your model?
> 
> As such, if we want to be consistent, the `multi_label` model should be post processed with a sigmoid, but that would not go with your models either.

That makes sense.....let me see what I can do about my models on my side, if it's as simple as just removing that lost line in my `forward()` method then we can clean this all up and add the sigmoid call to the `multi_label` code path and the current softmax call to the non-`multi_label` code path and call it a day :)

Bycob commented:
@cchadowitz-pf multi-label is now post processed by sigmoid, the rest is unchanged. Let me know if it's fine for you

cchadowitz-pf commented:
Cool thanks! I may not be able to get to it today but I'll test it out and let you know.

Bycob commented:
Hi @cchadowitz-pf, it would be better if you put the dlib update in another PR, so that we can merge this one.

cchadowitz-pf commented:
> Hi @cchadowitz-pf, it would be better if you put the dlib update in another PR, so that we can merge this one.

Sounds good - I've reverted the dlib changes here and moved them to a separate PR (https://github.com/jolibrain/deepdetect/pull/1452)  that should be merged **after** this one.
